File: backend/app/api/tools.py
import uuid
from fastapi import APIRouter, Depends, Query, UploadFile, File, Form
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.user import User
from app.schemas.tools import (
    ShortLinkCreate, ShortLinkResponse, PaginatedShortLinks,
    ImageRecordResponse, PaginatedImages,
)
from app.services.shortlink_service import (
    create_shortlink, get_user_shortlinks, delete_shortlink, shortlink_to_response,
)
from app.services.image_service import compress_and_save, get_user_images, image_record_to_response

router = APIRouter(dependencies=[Depends(get_current_user)])


# ---- Short Links ----
@router.post("/shortlinks", response_model=ShortLinkResponse, status_code=201)
async def create_link(
    data: ShortLinkCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    link = await create_shortlink(db, user.id, data.original_url)
    return shortlink_to_response(link)


@router.get("/shortlinks", response_model=PaginatedShortLinks)
async def list_links(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    items, total = await get_user_shortlinks(db, user.id, page=page, page_size=page_size)
    return PaginatedShortLinks(
        items=[shortlink_to_response(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.delete("/shortlinks/{link_id}", status_code=204)
async def delete_link(
    link_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await delete_shortlink(db, link_id, user.id)


# ---- Images ----
@router.post("/images/upload", response_model=ImageRecordResponse, status_code=201)
async def upload_image(
    file: UploadFile = File(...),
    quality: int = Form(80),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    record = await compress_and_save(db, user.id, file, quality)
    return image_record_to_response(record)


@router.get("/images", response_model=PaginatedImages)
async def list_images(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    items, total = await get_user_images(db, user.id, page=page, page_size=page_size)
    return PaginatedImages(
        items=[image_record_to_response(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )

File: frontend/src/api/tools.ts
import client from './client';
import type { ShortLinkItem, PaginatedShortLinks, ImageRecordItem, PaginatedImages } from '../types';

// ShortLinks
export async function createShortLink(original_url: string): Promise<ShortLinkItem> {
  const { data } = await client.post<ShortLinkItem>('/tools/shortlinks', { original_url });
  return data;
}
export async function fetchShortLinks(page = 1, page_size = 20): Promise<PaginatedShortLinks> {
  const { data } = await client.get<PaginatedShortLinks>('/tools/shortlinks', { params: { page, page_size } });
  return data;
}
export async function deleteShortLink(id: string): Promise<void> {
  await client.delete(`/tools/shortlinks/${id}`);
}

// Images
export async function uploadImage(file: File, quality = 80): Promise<ImageRecordItem> {
  const form = new FormData(); form.append('file', file); form.append('quality', String(quality));
  const { data } = await client.post<ImageRecordItem>('/tools/images/upload', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data;
}
export async function fetchImages(page = 1, page_size = 20): Promise<PaginatedImages> {
  const { data } = await client.get<PaginatedImages>('/tools/images', { params: { page, page_size } });
  return data;
}
